src/Views/VPNPage.vala

```diff
@@ -274,50 +218,123 @@ public class Network.VPNPage : Network.Widgets.Page {
+
+        var selected_row = vpn_list.get_selected_row () as VPNMenuItem;
+        if (selected_row != null) {
+            command = "nm-connection-editor --edit=" + selected_row.connection.get_uuid ();
+        }
+
         try {
-            network_manager.client.deactivate_connection (active_connection);
+            var appinfo = AppInfo.create_from_commandline (
+                command,
+                null,
+                AppInfoCreateFlags.NONE
+            );
+            appinfo.launch (null, null);
+        } catch (Error e) {
+            warning ("%s", e.message);
+        }
+    }
+
+    private bool create_connection () {
+        try {
+            var command = AppInfo.create_from_commandline ("nm-connection-editor --create --type=vpn", null, GLib.AppInfoCreateFlags.NONE);
+            command.launch (null, null);
+
         } catch (Error e) {
-            warning (e.message);
+            var dialog = new Granite.MessageDialog (
+                _("Failed to run Connection Editor"),
+                _("The program \"nm-connection-editor\" may not be installed."),
+                new ThemedIcon ("network-vpn"),
+                Gtk.ButtonsType.CLOSE
+            );
+            dialog.badge_icon = new ThemedIcon ("dialog-error");
+            dialog.show_error_details (e.message);
+            dialog.transient_for = (Gtk.Window) get_toplevel ();
+            dialog.run ();
+            dialog.destroy ();
+        }
+
+        return true;
+    }
+
+    private void delete_connection () {
+        var selected_row = vpn_list.get_selected_row () as VPNMenuItem;
+        if (selected_row != null && sel_row != null && sel_row == selected_row) {
+            try {
+                selected_row.connection.delete (null);
+            } catch (Error e) {
+                warning (e.message);
+                var dialog = new Granite.MessageDialog (
+                    _("Failed to remove VPN connection"),
+                    "",
+                    new ThemedIcon ("network-vpn"),
+                    Gtk.ButtonsType.CLOSE
+                );
+                dialog.badge_icon = new ThemedIcon ("dialog-error");
+                dialog.show_error_details (e.message);
+                dialog.transient_for = (Gtk.Window) get_toplevel ();
+                dialog.run ();
+                dialog.destroy ();
+            }
+        } else {
+            warning ("Row selection changed between operations. Cancelling removal of VPN.");
+            GLib.Source.remove (timeout_id);
+            timeout_id = 0;
+            sel_row.show ();
+        }
+
+    }
+
+    [CCode (instance_pos = -1)]
+    private int compare_rows (VPNMenuItem row1, VPNMenuItem row2) {
+        unowned NM.SettingConnection vpn_menu_item1 = row1.connection.get_setting_connection ();
+        unowned NM.SettingConnection vpn_menu_item2 = row2.connection.get_setting_connection ();
+
+        if (vpn_menu_item1.get_timestamp () > vpn_menu_item2.get_timestamp ()) {
```

Comment:
We should check if both `SettingConnection`s are not null here, since they can be.